File: listings/mflos.c
SM_OUTPUT(FloS1, Flooded);
SM_DATA(FloS1) int flooded, count;

SM_INPUT(FloS1, Sensor, data, len) {
  if (len > 0) {
    if (data[0] >= SATURATED) {
      flooded = 1;
    } else {
      flooded = count = 0;
      Flooded(&flooded, sizeof(flooded)); 
    }
  }
}
SM_INPUT(FloS1, Tick, data, len) {
  if (flooded && ++count > MAX) {
    Flooded(&flooded, sizeof(flooded)); 
  }
}

File: listings/mflos.rs
/* global variables definition omitted for brevity */

//@ sm_output(Flooded)

//@ sm_input
pub fn Sensor(data : &[u8]) {
  if data.len() > 0 {
    if data[0] >= SATURATED {
      flooded = true;
    } else {
      flooded = false;
      Flooded(&[0]); 
    }
  }
}
//@ sm_input
pub fn Tick(_data : &[u8]) {
  if flooded {
    count += 1;
    if count > MAX {
      Flooded(&[1]); 
    }
  }
}